Add branch protection doc

alvaroaleman opened the pull request:
Ref https://issues.redhat.com/browse/DPTP-1569

stevekuznetsov commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,33 @@
+
+## How do we set up branch protection?
+
+Branch protection is configured by a job that runs periodically every six hours. You can see
```
Right now you answer the question in the last part of this paragraph. Can we flip that and be more clear?

```
## How do we set up Branch Protection?

Branch Protection requires no manual set-up for repositories with pre-submit jobs. When a mandatory pre-submit job is added or removed from a repository, automation will ensure that the GitHub Branch Protection settings are updated as well. Branch Protection automation runs periodically, every six hours, so it may take some time for changes to jobs to be reflected. See the history of the job [here](link).
```

stevekuznetsov reviewed (commented):
We want documents on this site to be, as much as is feasible, the one-stop destination for help. While linking out is useful, ultimately we need to answer questions at this page and not another couple of clicks later.

alvaroaleman commented:
> We want documents on this site to be, as much as is feasible, the one-stop destination for help. While linking out is useful, ultimately we need to answer questions at this page and not another couple of clicks later.

I disagree. I am definitely not going to duplicate documentation from test-infra. We can try to find a better way than linking, but dupicating doesn't make any sense.

alvaroaleman commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,33 @@
+---
```
But?

stevekuznetsov replied:
But we don't have a better place to put it ... :(

alvaroaleman commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,33 @@
+---
+## What is branch protection?
+
+Branch protection is a repository setting that enforces certain rules when trying to merge to a protected branch. See
```
Have you followed that link? Its more confusing than anything. Branch protection is setting up requirements to allow merging, and that is in the doc.

petr-muller commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,33 @@
+
+Yes, these settings are configured here: https://github.com/openshift/release/blob/master/core-services/prow/02_config/_config.yaml
+
+An overview of all available settings and a more detailled explanation can be found in the
```
Sure. As a followup?

stevekuznetsov replied:
YEs

petr-muller reviewed (commented):
This is not really a comment that I could assign to specific paragraph or sentence here, more like general thoughts on the content.

I think that people are not actually that interested in branch protection as a mechanism (the ones that are, we should send to the GH doc ASAP). They are probably not that interested in branchprotector as a prow component either, our users are more interested in how BP applies to them, so IMO we should focus not only on what's possible with BP, but also about major specifics of our deployment of it.

What I'd like to convey to the user is:

- Branch protection is kind of a "second" layer (first one being Tide) that infuences which PRs are mergeable and which are not. This layer is unfortunately the one that is more prominently visible in GH PRs (with the `required` stamps on jobs). We infer the BP settings from Prow job settings (this is important).
- I would definitely point out the "default" state more prominently: If the repo has no custom BP configuration, we wil set BP on exactly those branches that have Prow jobs. When you set Prow jobs, we *will* set BP. We should also explain which jobs will be marked required and which not - even Tide-required jobs will not be set as required in BP, if they are not always executed.
- I would add paragraphs instructing on how repos can deviate from this default state (needing specific BP config, disabling BP, not touching BP)
- The fact that our merge bot needs admin rights to set BP

I think the rest can be solved with links to other documentation (GH, Prow).

alvaroaleman commented:
@petr-muller 

> Branch protection is kind of a "second" layer (first one being Tide) that infuences which PRs are mergeable and which are not. This layer is unfortunately the one that is more prominently visible in GH PRs (with the required stamps on jobs). We infer the BP settings from Prow job settings (this is important).

Is it? To the best of my knowledge the only thing branch protection does is telling users what jobs are always mandatory. The authority over "Can I merge this" is the tide status context, because the jobs status contexts don't contain the info if the job was tested against the latest revision of the base. We only happen to not support protecting the tide status context (ref https://github.com/kubernetes/test-infra/issues/16472)

>  would definitely point out the "default" state more prominently: If the repo has no custom BP configuration, we wil set BP on exactly those branches that have Prow jobs. When you set Prow jobs, we will set BP. We should also explain which jobs will be marked required and which not - even Tide-required jobs will not be set as required in BP, if they are not always executed.

I've added something in the `How do we set up branch protection?`, do you think that is sufficient?

> I would add paragraphs instructing on how repos can deviate from this default state (needing specific BP config, disabling BP, not touching BP)

I've added something in the `Is it possible to disable the branch protection for my repository or require jobs that are not managed by Prow?` section, do you think that is sufficient?

> The fact that our merge bot needs admin rights to set BP

Added that.

stevekuznetsov commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,70 @@
+date: 2020-10-05T11:14:39-04:00
+draft: false
+---
+## What is branch protection?
```
Nit: can we title case headers?

stevekuznetsov commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,70 @@
+This status is generated by GitHub and means that Prow didn't report the job as running yet.
+
+Additionally, the branch protection settings are interpreted by the merge automation
+[tide](https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/README.md) in order to [determine if jobs that
```
```suggestion
[`tide`](https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/README.md) in order to [determine if jobs that
```

stevekuznetsov commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,70 @@
+
+## How do we set up branch protection?
+
+Branch Protection requires no manual set-up for repositories with pre-submit jobs. When a mandatory presubmit job is
```
nit: this has pre-submit and presubmit, choose one?

alvaroaleman commented:
/retest

stevekuznetsov commented on content/docs/architecture/branch-protection.md:
```diff
@@ -0,0 +1,70 @@
+![Required Status context](/branchprotection_sample_required_job.png)
+
+
+If you open a new Pull Request to a branch that has required job, these jobs will initially be marked as
```
```suggestion
If you open a new Pull Request to a branch that has required jobs, these jobs will initially be marked as
```

stevekuznetsov reviewed (commented):
Sorry, missed one typo, otherwise LGTM

stevekuznetsov commented:
/lgtm
